File: 백준/Gold/18427. 함께 블록 쌓기/함께 블록 쌓기.java
import java.util.*;
import java.io.*;

public class Main {
	public static void main(String[] args) throws IOException{
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		st = new StringTokenizer(br.readLine());
		n = Integer.parseInt(st.nextToken());
		m = Integer.parseInt(st.nextToken());
		h = Integer.parseInt(st.nextToken());
		
		arr = new int[n+1][m];
		for(int i = 1; i<=n; i++) {
			st = new StringTokenizer(br.readLine());
			int size = st.countTokens();
			for(int j = 0; j < size;j++) {
				arr[i][j] = Integer.parseInt(st.nextToken());
			}
//			System.out.println(Arrays.toString(arr[i]));
		}
		
		dp = new int[n+1][h+1];
		for(int row[] : dp) {
			Arrays.fill(row, -1);
		}
		
		ans = solve(1,0);
		System.out.println(ans);
	}
	
	static int dp[][];
	
	static int solve(int cur, int sum) {
		if(sum == h)
			return 1;
		if(cur > n || sum > h)
			return 0;
        
		if(dp[cur][sum] != -1)
			return dp[cur][sum];
		
		int ret = 0;
		for(int i=0; i< m; i++) {
			if(arr[cur][i] != 0)
				ret = (ret + solve(cur + 1, sum + arr[cur][i])) % mod;;
				
		}
		ret = (ret +  solve(cur + 1, sum)) % mod;
		return dp[cur][sum] = ret;
	}
	
//	static void solve(int cur, int sum) {
//		if(sum == h) {
//			ans++;
//			return;
//		}
//		
//		
//		if(cur == n + 1)
//			return;
//		
//		if(sum > h)
//			return;
//		
//		for(int i = 0; i < m ;i++) {
//			if(arr[cur][i] != 0) {
//				solve(cur + 1, sum + arr[cur][i]);
//			}
//		}
//		solve(cur + 1, sum);
//	}
	
	static int mod = 10007;
	static int ans;
	static int n,m,h;
	static StringTokenizer st;
	static int arr[][];
}

File: 백준/Gold/18427. 함께 블록 쌓기/함께 블록 쌓기.py
from collections import defaultdict

def solve(cur, sum):
    # print(f"cur: {cur}, sum: {sum}")
    if sum == H:
        return 1

    if sum > H:
        return 0

    if cur == N:
        return 0

    if (cur,sum) in dp:
        return dp[(cur,sum)]

    ret = solve(cur + 1, sum)

    for block in blocks[cur]:
        ret = (ret + solve(cur+1, sum + block)) % MOD

    dp[(cur,sum)] = ret
    return ret


if __name__ == "__main__":
    N,M,H = map(int,input().split())
    blocks = [[] for _ in range(N)]
    MOD = 10007
    for i in range(N):
        blocks[i] = list(map(int,input().split()))

    dp = defaultdict(int)

    print(solve(0, 0))